### ISSUE TITLE
ELF-File not reconized 

### ISSUE THREAD

**Issue by nexoscp** [ferrous-systems/nrfdfu-rs#14]
Build the "blinky" example from https://github.com/nrf-rs/adafruit-nrf52-bluefruit-le

Ensured i got a elf file: 
**file target/thumbv7em-none-eabihf/release/examples/blinky**
**target/thumbv7em-none-eabihf/release/examples/blinky: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, with debug_info, not stripped**

Tried to use nrfdfu and got this error:
**nrfdfu target/thumbv7em-none-eabihf/release/examples/blinky**
**error: no loadable program segments found; ensure that the linker is invoked correctly (passing the linker script)**

Not sure where to go from here.

**Comment by japaric** [comment]
that repository appears to be incomplete. if you build the code as is, you end with an empty ELF file
``` console
$ cargo build --target thumbv7em-none-eabihf --example blinky --release

$ size -Ax target/thumbv7em-none-eabihf/release/examples/blinky
target/thumbv7em-none-eabihf/release/examples/blinky  :
section              size   addr
.ARM.attributes      0x3a    0x0
(..)
```

no `.text` section means that the ELF contains no instructions. As `nrfdfu` indicates, it appears that no linker script is being passed to the linker.

It seems that the `.cargo/config.toml` file is missing. Try adding this:

``` console
$ mkdir .cargo
$ cat >> .cargo/config.toml << STOP
[build]
target = "thumbv7em-none-eabihf"

[target.thumbv7em-none-eabihf]
rustflags = [
  "-C", "link-arg=-Tlink.x",
]
STOP
```

and re-build the firmware.

``` console
$ cargo clean

$ cargo build --example blinky --release

$ size -Ax target/thumbv7em-none-eabihf/release/examples/blinky
target/thumbv7em-none-eabihf/release/examples/blinky  :
section              size         addr
.vector_table        0xdc      0x26000
.text              0x3154      0x260dc
.rodata             0xee8      0x29230
.data                 0x0   0x20003400
.bss                  0x4   0x20003400
(..)
```

The new version has the required sections.

**Comment by nexoscp** [comment]
Thanks for the help! Stupid of me.
 But  now i get "error: firmware starts at address 0x26000, expected 0x1000". Seams to come from here:
https://github.com/ferrous-systems/nrfdfu-rs/blob/e9700a580e046b68f2be49627ebe4f42c08c8a5a/src/elf.rs#L88
Moving the address to 0x1000 in memory.x  helps. But don't know how to get it together with a "SoftDevice" blob.

**Comment by japaric** [comment]
>  Seams to come from here:

this tool was created for use with the nrf-dongle as part of [our training material](https://github.com/ferrous-systems/embedded-trainings-2020). the nrf-dongle dev board has a bootloader in the 0x0000 - 0x1000 address range. that check there is to ensure one does not overwrite the bootloader which would render the bootloader un-flashable (not quite "brick it" as one re-install the firmware using a separate hardware debugger but that's cumbersome / annoying).

in your case, if you are certain that the firmware starts at 0x26000 you could modify the value in https://github.com/ferrous-systems/nrfdfu-rs/blob/895747421db3b3a6979d95312aae9af6240af317/src/elf.rs#L87

to 0x26000 or even remove the entire check. I'm not sure if `nrfdfu` has other hard-coded values that are specific to the nrf-dongle. @jonas-schievink may be able to comment on that.

**Comment by jonas-schievink** [comment]
Hmm, I would expect the bootloader to refuse overwriting itself, but I haven't tested that. I think that check was mostly added as a precaution for flashing incorrectly linked binaries, not for preventing bricks. If it really does work correctly with an address of 0x26000 we can probably just remove the check. Could you test this @nexoscp?

**Comment by nexoscp** [comment]
Would `if addr < 0x1000` fix my problem and keep the check still working?

**Comment by jonas-schievink** [comment]
That should work too, yeah

**Comment by nexoscp** [comment]
Have a look: https://github.com/ferrous-systems/nrfdfu-rs/pull/15 Worked for the blinky example.

**Comment by japaric** [comment]
released v0.1.3 which includes PR #15. @nexoscp thanks again for working on this

**Comment by nexoscp** [comment]
Thanks for helping!
